File: java/Sqrt.java
/*
 * Filename: Sqrt.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 19 May 2015 02:09:01 PM EDT
 */

public class Solution {
	public int mySqrt(int x) {
		double res = Math.pow(Math.E, Math.log(x) / 2.0);
		int resInt = (int) res;

		if (Math.abs(resInt + 1 - res) < 0.000001) {
			return resInt + 1;
		} else {
			return resInt;
		}
	}
}

File: cpp/sqrt.cpp
/*
 * Filename: sqrt.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Wed 27 Jan 2016 08:50:31 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution {
	public:
		int mySqrt(int x) {
			if (x <= 0)
			{
				return 0;
			}

			int base = 0;
			while (true)
			{
				int i = 1;
				int val = base + i;

				while ((double) val * (double) val <= (double) x)
				{
					if (val * val == x)
					{
						return val;
					}

					i *= 2;
					val = base + i;
				}

				if (i/2 == 0)
				{
					break;
				}

				base += i/2;
			}

			return base;
		}
};
